libs/contrib/Data/Vect/Views/Extra.idr

```diff
@@ -0,0 +1,66 @@
+||| Additional views for Vect
+module Data.Vect.Views.Extra
+
+import Control.WellFounded
+import Data.Vect
+import Data.Nat
+
+
+||| View for splitting a vector in half, non-recursively
+public export
+data Split : Vect n a -> Type where
+    SplitNil : Split []
+    SplitOne : Split [x]
+    ||| two non-empty parts
+    SplitPair : {n : Nat} -> {m : Nat} ->
+                {x, y : a} -> {xs : Vect n a} -> {ys : Vect m a} ->
+                Split (x :: xs ++ y :: ys)
+
+total
+splitHelp : {n: Nat} -> (head : a) -> (zs : Vect n a) ->
+            Nat -> Split (head :: zs)
+splitHelp head [] _ = SplitOne
+splitHelp head (z :: zs) 0 = SplitPair {xs = []} {ys = zs}
+splitHelp head (z :: zs) 1 = SplitPair {xs = []} {ys = zs}
+splitHelp head (z :: zs) (S (S k))
+  = case splitHelp head zs k of
+         SplitOne => SplitPair {xs = []} {ys = []}
+         SplitPair {xs} {ys} => SplitPair {xs = z :: xs} {ys = ys}
+
+||| Covering function for the `Split` view
+||| Constructs the view in linear time
+export total
+split : {n: Nat} -> (xs : Vect n a) -> Split xs
+split [] = SplitNil
+split {n = S k} (x :: xs) = splitHelp x xs k
+
+||| View for splitting a vector in half, recursively
+|||
+||| This allows us to define recursive functions which repeatedly split vectors
+||| in half, with base cases for the empty and singleton lists.
+public export
+data SplitRec : Vect k a -> Type where
+  SplitRecNil : SplitRec []
+  SplitRecOne : SplitRec [x]
+  SplitRecPair : {n : Nat} -> {m : Nat} -> {xs : Vect n a} -> {ys : Vect m a} ->
```

Comment:
Shouldn't we have `{xs : Vect (S n) a} -> {ys : Vect (S m) a} ->` here?
Otherwise the split is not necessarily unique: `SplitRecPair SplitNil SplitNil`
and `SplitNil` are 2 valid splits of the same empty vector.